# Business Problem

KBO Marketing Company LLC is a marketing company.  KBO Marketing has been retained by Apple and Google to perform the following:

- understand product sentiment analysis - in particular, whether or not the product sentiment is positive or not positive;
- and proactively respond to feedback that is less than positive via social media  

# Data Understanding

The data for examing the aforementioned problem comes from the following source: [Brands and Product Emotions](https://data.world/crowdflower/brands-and-product-emotions) 

Before beginning to create a model, I want to examine and become familiar with the dataset. I will conduct exploratory data analysis in order to understand the dataset attributes, which includes, but not limited to the following:

1. Number of Columns
2. Number of Rows
3. Column Names
4. Format of the data in each column

In [1]:
# Importing Libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from itertools import combinations

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import recall_score, precision_score, accuracy_score, f1_score, confusion_matrix, ConfusionMatrixDisplay, roc_curve, auc
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier 
from sklearn import tree

import nltk
import re
nltk.download('punkt')
from nltk.tokenize import regexp_tokenize, word_tokenize, RegexpTokenizer
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer


[nltk_data] Downloading package punkt to C:\Users\16467/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\16467/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [2]:
# Reading file

df = pd.read_csv('tweet_product_company.csv')

In [3]:
# Examining first 10 rows

df.head(10)

tweet_text  \
0  .@wesley83 I have a 3G iPhone. After 3 hrs twe...   
1  @jessedee Know about @fludapp ? Awesome iPad/i...   
2  @swonderlin Can not wait for #iPad 2 also. The...   
3  @sxsw I hope this year's festival isn't as cra...   
4  @sxtxstate great stuff on Fri #SXSW: Marissa M...   
5  @teachntech00 New iPad Apps For #SpeechTherapy...   
6                                                NaN   
7  #SXSW is just starting, #CTIA is around the co...   
8  Beautifully smart and simple idea RT @madebyma...   
9  Counting down the days to #sxsw plus strong Ca...   

  emotion_in_tweet_is_directed_at  \
0                          iPhone   
1              iPad or iPhone App   
2                            iPad   
3              iPad or iPhone App   
4                          Google   
5                             NaN   
6                             NaN   
7                         Android   
8              iPad or iPhone App   
9                           Apple   

  is_there_an_emotion_directed_at_a_brand_or_product  
0                                   Negative emotion  
1                                   Positive emotion  
2                                   Positive emotion  
3                                   Negative emotion  
4                                   Positive emotion  
5                 No emotion toward brand or product  
6                 No emotion toward brand or product  
7                                   Positive emotion  
8                                   Positive emotion  
9                                   Positive emotion

In [4]:
# Examining last 10 rows

df.tail(10)

tweet_text  \
8711  Google says the future is all around you! (ie,...   
8712  Google says the future is location, location, ...   
8713  I've always used Camera+ for my iPhone b/c it ...   
8714  Google says: want to give a lightning talk to ...   
8715  @mention Yup, but I don't have a third app yet...   
8716                      Ipad everywhere. #SXSW {link}   
8717  Wave, buzz... RT @mention We interrupt your re...   
8718  Google's Zeiger, a physician never reported po...   
8719  Some Verizon iPhone customers complained their...   
8720  �ϡ�����_��ʋ�΋�ҋ�������⋁_��������_���RT @mentio...   

     emotion_in_tweet_is_directed_at  \
8711                             NaN   
8712                             NaN   
8713              iPad or iPhone App   
8714                             NaN   
8715                             NaN   
8716                            iPad   
8717                             NaN   
8718                             NaN   
8719                             NaN   
8720                             NaN   

     is_there_an_emotion_directed_at_a_brand_or_product  
8711                 No emotion toward brand or product  
8712                 No emotion toward brand or product  
8713                                   Positive emotion  
8714                 No emotion toward brand or product  
8715                 No emotion toward brand or product  
8716                                   Positive emotion  
8717                 No emotion toward brand or product  
8718                 No emotion toward brand or product  
8719                 No emotion toward brand or product  
8720                 No emotion toward brand or product

In [5]:
# Getting column info

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8721 entries, 0 to 8720
Data columns (total 3 columns):
 #   Column                                              Non-Null Count  Dtype 
---  ------                                              --------------  ----- 
 0   tweet_text                                          8720 non-null   object
 1   emotion_in_tweet_is_directed_at                     3169 non-null   object
 2   is_there_an_emotion_directed_at_a_brand_or_product  8721 non-null   object
dtypes: object(3)
memory usage: 204.5+ KB


## Data Overview

I created a Pandas DataFrame from the csv file.  The DataFrame contains 8,721 rows of data.  The DataFrame contains the following columns:

- *df['tweet_text']* - this is the sentiment provided via tweet
- *df['emotion_in_tweet_is_directed_at']* - this is the brand, or product, in which the tweet is targeted
- *df['is_there_an_emotion_directed_at_a_brand_or_product']* - this is the emotional category of the tweet

All of the columns are in string format.

In [6]:
# Obtaining number of unique values for the following column - "df['emotion_in_tweet_is_directed_at']"

df['emotion_in_tweet_is_directed_at'].nunique()

9

In [7]:
# Obtaining Value Counts for the following column - "df['emotion_in_tweet_is_directed_at']"

df['emotion_in_tweet_is_directed_at'].value_counts()

iPad                               910
Apple                              640
iPad or iPhone App                 451
Google                             412
iPhone                             288
Other Google product or service    282
Android App                         78
Android                             74
Other Apple product or service      34
Name: emotion_in_tweet_is_directed_at, dtype: int64

## Observations | Brand or Product

I utilized the following code - *df['emotion_in_tweet_is_directed_at'].value_counts()* - to understand the different brands and products that are captured by the respective column.  The different brands and products are the following: 

- iPad
- Apple
- iPad or iPhone App
- Google
- iPhone
- Other Google product or service
- Android App
- Android
- Other Apple product or service

In [10]:
# Obtaining number of unique values for the following column - "df['is_there_an_emotion_directed_at_a_brand_or_product']"

df['is_there_an_emotion_directed_at_a_brand_or_product'].nunique()

4

In [9]:
# Obtaining Value Counts for the following column - "df['is_there_an_emotion_directed_at_a_brand_or_product']"

df['is_there_an_emotion_directed_at_a_brand_or_product'].value_counts()

No emotion toward brand or product    5156
Positive emotion                      2869
Negative emotion                       545
I can't tell                           151
Name: is_there_an_emotion_directed_at_a_brand_or_product, dtype: int64

## Observations | Brand or Product Sentiment

I utilized the following code - *df['is_there_an_emotion_directed_at_a_brand_or_product'].value_counts()* - to understand the brand and product sentiment captured by the respective column.  There are a total of four different brand and product sentiments, which are the following:

- No emotion toward brand or product
- Positive emotion
- Negative emotion
- I can't tell 

In [11]:
# Examining missing values in each column

df.isna().sum()

tweet_text                                               1
emotion_in_tweet_is_directed_at                       5552
is_there_an_emotion_directed_at_a_brand_or_product       0
dtype: int64

## Observations | Missing Values

I utilized the following code - *df.isna().sum()* - to understand how many missing values are within each column.  There is a total of 5,552 missing values in the *df['emotion_in_tweet_is_directed_at']* column.

The other two columns - *df['tweet_text']* and *df['is_there_an_emotion_directed_at_a_brand_or_product']* - did not have missing values.

In [12]:
# Checking for duplicates

df.duplicated().sum()

22

In [13]:
# Creating a dataframe for the duplicates

duplicates_all = df[df.duplicated(keep=False)]

In [14]:
# Examining the dataframe created for the duplicates

duplicates_all

tweet_text  \
7     #SXSW is just starting, #CTIA is around the co...   
9     Counting down the days to #sxsw plus strong Ca...   
17    I just noticed DST is coming this weekend. How...   
20    Need to buy an iPad2 while I'm in Austin at #s...   
21    Oh. My. God. The #SXSW app for iPad is pure, u...   
24    Really enjoying the changes in Gowalla 3.0 for...   
455      Before It Even Begins, Apple Wins #SXSW {link}   
457      Before It Even Begins, Apple Wins #SXSW {link}   
750   Google to Launch Major New Social Network Call...   
752   Google to Launch Major New Social Network Call...   
2136  Marissa Mayer: Google Will Connect the Digital...   
2138  Marissa Mayer: Google Will Connect the Digital...   
2437  Counting down the days to #sxsw plus strong Ca...   
3759  Really enjoying the changes in Gowalla 3.0 for...   
3771  #SXSW is just starting, #CTIA is around the co...   
4669  Oh. My. God. The #SXSW app for iPad is pure, u...   
5105  RT @mention ��� GO BEYOND BORDERS! ��_ {link} ...   
5107  RT @mention ��� GO BEYOND BORDERS! ��_ {link} ...   
5108  RT @mention ��� Happy Woman's Day! Make love, ...   
5110  RT @mention ��� Happy Woman's Day! Make love, ...   
5648  RT @mention Google to Launch Major New Social ...   
5649  RT @mention Google to Launch Major New Social ...   
5650  RT @mention Google to Launch Major New Social ...   
5651  RT @mention Google to Launch Major New Social ...   
5652  RT @mention Google to Launch Major New Social ...   
5653  RT @mention Google to Launch Major New Social ...   
5654  RT @mention Google to Launch Major New Social ...   
6061  RT @mention Marissa Mayer: Google Will Connect...   
6063  RT @mention Marissa Mayer: Google Will Connect...   
6064  RT @mention Marissa Mayer: Google Will Connect...   
6065  RT @mention Marissa Mayer: Google Will Connect...   
6066  RT @mention Marissa Mayer: Google Will Connect...   
6067  RT @mention Marissa Mayer: Google Will Connect...   
6068  RT @mention Marissa Mayer: Google Will Connect...   
6069  RT @mention Marissa Mayer: Google Will Connect...   
6313  RT @mention RT @mention Google to Launch Major...   
6315  RT @mention RT @mention Google to Launch Major...   
8146  I just noticed DST is coming this weekend. How...   
8394  Need to buy an iPad2 while I'm in Austin at #s...   

     emotion_in_tweet_is_directed_at  \
7                            Android   
9                              Apple   
17                            iPhone   
20                              iPad   
21                iPad or iPhone App   
24                       Android App   
455                            Apple   
457                            Apple   
750                              NaN   
752                              NaN   
2136                             NaN   
2138                             NaN   
2437                           Apple   
3759                     Android App   
3771                         Android   
4669              iPad or iPhone App   
5105                             NaN   
5107                             NaN   
5108                             NaN   
5110                             NaN   
5648                             NaN   
5649                             NaN   
5650                             NaN   
5651                             NaN   
5652                             NaN   
5653                             NaN   
5654                             NaN   
6061                          Google   
6063                             NaN   
6064                             NaN   
6065                          Google   
6066                             NaN   
6067                          Google   
6068                             NaN   
6069                             NaN   
6313                             NaN   
6315                             NaN   
8146                          iPhone   
8394                            iPad   

     is_there_an_emotion_directed_at_a_brand_or_product  
7                        

## Observations | Duplicate Rows

I utilized the following code - *df.duplicated().sum()* - to understand how many duplicates are present in the dataset.  There is a total of 22 duplicate rows.

# Data Preparation

# Modeling

# Overall Conclusions and Recommendations